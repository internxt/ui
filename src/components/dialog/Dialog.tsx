import { useEffect } from 'react';
import { useState } from 'react';
import { Button } from '../button/Button';

export default function Dialog({
  isOpen,
  onClose,
  onPrimaryAction,
  onSecondaryAction,
  title,
  subtitle,
  primaryAction,
  secondaryAction,
  primaryActionColor,
  isLoading,
}: {
  isOpen: boolean;
  onClose: () => void;
  onPrimaryAction: () => void;
  onSecondaryAction: () => void;
  title: string;
  subtitle: string;
  primaryAction: string;
  secondaryAction: string;
  primaryActionColor: 'primary' | 'danger';
  isLoading?: boolean;
}): JSX.Element {
  const [isVisible, setIsVisible] = useState(isOpen);
  const [transitionOpacity, setTransitionOpacity] = useState<string>('opacity-0');
  const [transitionScale, setTransitionScale] = useState<string>('scale-95');

  useEffect(() => {
    if (isOpen) {
      const timeout = setTimeout(() => {
        setTransitionOpacity('opacity-100');
        setTransitionScale('scale-100');
      }, 10);
      setIsVisible(true);
      return () => clearTimeout(timeout);
    } else {
      setTransitionOpacity('opacity-0');
      setTransitionScale('scale-95');
      const timeout = setTimeout(() => {
        setIsVisible(false);
      }, 150);
      return () => clearTimeout(timeout);
    }
  }, [isOpen]);

  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        onClose();
      }
    };

    if (isOpen) {
      window.addEventListener('keydown', handleKeyDown);
    }

    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [isOpen, onClose]);

  return (
    <>
      {isVisible && (
        <div className={`fixed inset-0 z-50 ${isOpen ? '' : 'pointer-events-none'}`}>
          <div
            className={`absolute inset-0 bg-gray-100/50 transition-opacity
             duration-150 dark:bg-black/75
              ${transitionOpacity}
            `}
            onClick={onClose}
            data-testid="dialog-overlay"
          ></div>

          <div
            className={`absolute
              left-1/2
              top-1/2
              w-full
              max-w-sm
              -translate-x-1/2
              -translate-y-1/2
              transform rounded-2xl
              bg-surface p-5
              transition-all
              duration-150
              dark:bg-gray-1
              ${transitionScale} 
              ${transitionOpacity}`}
          >
            <div className="flex flex-col space-y-2">
              <p className="text-2xl font-medium text-gray-100">{title}</p>
              <p className="text-gray-60">{subtitle}</p>
            </div>

            <div className="mt-5 flex justify-end space-x-2">
              <Button variant="secondary" onClick={onSecondaryAction} disabled={isLoading}>
                {secondaryAction}
              </Button>
              <Button
                onClick={onPrimaryAction}
                loading={isLoading}
                variant={primaryActionColor === 'primary' ? 'primary' : 'destructive'}
              >
                {primaryAction}
              </Button>
            </div>
          </div>
        </div>
      )}
    </>
  );
}
