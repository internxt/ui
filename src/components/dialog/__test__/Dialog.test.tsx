import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, vi, expect, afterEach } from 'vitest';
import Dialog from '../Dialog';

describe('Dialog', () => {
  const defaultProps = {
    isOpen: true,
    onClose: vi.fn(),
    onPrimaryAction: vi.fn(),
    onSecondaryAction: vi.fn(),
    title: 'Dialog Title',
    subtitle: 'Dialog Subtitle',
    primaryAction: 'Confirm',
    secondaryAction: 'Cancel',
    primaryActionColor: 'primary' as const,
    isLoading: false,
  };

  const renderDialog = (props = {}) => render(<Dialog {...defaultProps} {...props} />);

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should match snapshot', () => {
    const { container } = renderDialog();
    expect(container).toMatchSnapshot();
  });

  it('renders dialog content when open', () => {
    const { getByText } = renderDialog();
    expect(getByText(defaultProps.title)).toBeInTheDocument();
    expect(getByText(defaultProps.subtitle)).toBeInTheDocument();
  });

  it('does not render dialog content when closed', () => {
    const { queryByText } = renderDialog({ isOpen: false });
    expect(queryByText(defaultProps.title)).not.toBeInTheDocument();
  });

  it('calls onClose when clicking outside the dialog', () => {
    const { getByTestId } = renderDialog();
    fireEvent.click(getByTestId('dialog-overlay'));
    expect(defaultProps.onClose).toHaveBeenCalled();
  });

  it('calls onClose when pressing Escape', () => {
    renderDialog();
    fireEvent.keyDown(window, { key: 'Escape' });
    expect(defaultProps.onClose).toHaveBeenCalled();
  });

  it('calls onPrimaryAction when clicking the primary button', () => {
    const { getByText } = renderDialog();
    fireEvent.click(getByText(defaultProps.primaryAction));
    expect(defaultProps.onPrimaryAction).toHaveBeenCalled();
  });

  it('calls onSecondaryAction when clicking the secondary button', () => {
    const { getByText } = renderDialog();
    fireEvent.click(getByText(defaultProps.secondaryAction));
    expect(defaultProps.onSecondaryAction).toHaveBeenCalled();
  });

  it('disables buttons when isLoading is true', () => {
    const { getAllByRole } = renderDialog({ isLoading: true });
    const elements = getAllByRole('button');
    screen.debug();
    expect(elements[0]).toBeDisabled();
    expect(elements[1]).toBeDisabled();
  });

  it('applies correct classes for primaryActionColor', () => {
    const { getAllByRole } = renderDialog({ primaryActionColor: 'danger' });
    const elements = getAllByRole('button');
    expect(elements[1]).toHaveClass('bg-red', 'active:bg-red-dark', 'text-white', 'shadow-sm');
  });
});
