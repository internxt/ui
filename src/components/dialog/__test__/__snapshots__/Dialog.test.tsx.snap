// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dialog > should match snapshot 1`] = `
<div>
  <div
    class="fixed inset-0 z-50 "
  >
    <div
      class="absolute inset-0 bg-gray-100/50 transition-opacity
             duration-150 dark:bg-black/75
              opacity-0
            "
      data-testid="dialog-overlay"
    />
    <div
      class="absolute
              left-1/2
              top-1/2
              w-full
              max-w-sm
              -translate-x-1/2
              -translate-y-1/2
              transform rounded-2xl
              bg-surface p-5
              transition-all
              duration-150
              dark:bg-gray-1
              scale-95 
              opacity-0"
    >
      <div
        class="flex flex-col space-y-2"
      >
        <p
          class="text-2xl font-medium text-gray-100"
        >
          Dialog Title
        </p>
        <p
          class="text-gray-60"
        >
          Dialog Subtitle
        </p>
      </div>
      <div
        class="mt-5 flex justify-end space-x-2"
      >
        <button
          class="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-surface dark:bg-gray-5 border border-gray-10 hover:border-gray-20 active:bg-gray-1 dark:active:bg-gray-10 text-gray-80 shadow-sm "
          type="button"
        >
          <div
            class="flex items-center justify-center space-x-2"
          >
            Cancel
          </div>
        </button>
        <button
          class="h-10 px-5 relative flex shrink-0 select-none flex-row items-center justify-center space-x-2 
      whitespace-nowrap rounded-lg text-base font-medium outline-none ring-2 ring-primary/0 
      ring-offset-2 ring-offset-transparent transition-all duration-100 ease-in-out 
      focus-visible:ring-primary/50 bg-primary active:bg-primary-dark text-white shadow-sm "
          type="button"
        >
          <div
            class="flex items-center justify-center space-x-2"
          >
            Confirm
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
